File: hayaati_alarm/hayaati_alarm/src/flutter/hayaati_alarm/lib/src/hayaati_alarm.dart
// hayaati_alarm.dart (src/)
//
// Implémentation du Service : reçoit les appels invoke_method() venant de
// Python et pilote le paquet Flutter 'alarm' (https://pub.dev/packages/alarm),
// activement maintenu, API vérifiée en date d'aujourd'hui sur sa page pub.dev.
//
// ── Journal des corrections (30/08/2026) ────────────────────────────
// v1.1 :
//  • fadeDuration et warningNotificationOnKill sont désormais lus dans
//    args et transmis à VolumeSettings / AlarmSettings, pour suivre
//    l'évolution du côté Python (hayaati_alarm.py v1.1).
//  • Le cas "isAlarmActive" a été retiré : il n'était de toute façon
//    jamais atteint avec succès côté Python, et son rôle (éviter les
//    doublons) est déjà couvert par cancel_all_prayer_alarms() avant
//    chaque reschedule_all().
//  • bypassDnd n'a jamais existé dans ce fichier ni dans l'API connue
//    du paquet alarm : rien à faire ici, juste ne pas l'ajouter côté
//    Python sans vérification préalable sur pub.dev.
//
// ── Journal des corrections (08/09/2026) ────────────────────────────
// v1.2 :
//  • Ajout de _ensureAudioFileExtracted(). Le paquet 'alarm' échouait
//    silencieusement à extraire lui-même l'asset audio quand celui-ci
//    vit dans un package tiers (hayaati_alarm) plutôt que dans l'app
//    principale : la clé d'asset qu'il résout en interne ne correspond
//    pas à la clé réelle dans flutter_assets. Confirmé en inspectant
//    l'APK décompressé : le fichier se trouve à
//    assets/flutter_assets/packages/hayaati_alarm/assets/sounds/adhan.mp3
//    On extrait donc l'asset nous-mêmes vers un fichier réel sur disque
//    avant d'appeler Alarm.set(), et on transmet ce chemin de fichier
//    plutôt que la clé d'asset brute.
//  • args["assetAudioPath"] envoyé par Python doit maintenant contenir
//    la clé d'asset complète avec sous-dossier, ex. :
//    "packages/hayaati_alarm/assets/sounds/adhan.mp3"
//
// ⚠️ À VÉRIFIER avant de lancer un build complet sur ta machine :
// 1. Que 'path_provider' figure bien dans pubspec.yaml de ce package
//    (hayaati_alarm/pubspec.yaml). Si absent : flutter pub add path_provider
// 2. Le nom exact du champ "warningNotificationOnKill" sur AlarmSettings,
//    et la signature de VolumeSettings.fade(...), peuvent varier d'une
//    version à l'autre du paquet 'alarm'. Lance d'abord :
//        cd hayaati_alarm/src/flutter/hayaati_alarm
//        flutter analyze
//    C'est net plus rapide qu'un build APK complet et ça suffit à repérer
//    un nom de champ qui aurait changé entre deux versions du paquet.
//
// ⚠️ Les noms exacts addInvokeMethodListener / removeInvokeMethodListener
// n'ont pas été modifiés : puisque l'app compile et s'installe déjà
// avec ce fichier, ces méthodes existent bien dans cette version de Flet.

// ⚠️ 09/09/2026 : import 'dart:io' retiré. `flutter analyze` signale
// (niveau info, pas une erreur) qu'il est redondant : package:flet/flet.dart
// réexporte déjà File, seule classe de dart:io utilisée dans ce fichier
// (_ensureAudioFileExtracted). Aucun changement de comportement.
import 'package:flutter/foundation.dart' show debugPrint;
import 'package:flutter/services.dart' show rootBundle;
import 'package:flet/flet.dart';
import 'package:alarm/alarm.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:path_provider/path_provider.dart';

class HayaatiAlarmService extends FletService {
  HayaatiAlarmService({required super.control});

  @override
  void init() {
    super.init();
    // ⚠️ DIAGNOSTIC 02/09/2026 : trace non filtrée, pour confirmer que le
    // service est bien créé une seule fois et n'est pas détruit puis
    // recréé entre l'enregistrement du listener et l'appel de Python.
    debugPrint(
      '[HayaatiAlarm][DIAG] init() — instance=${identityHashCode(this)} '
      'control.id=${control.id}',
    );
    Alarm.init().then((_) {
      debugPrint('[HayaatiAlarm][DIAG] Alarm.init() terminé avec succès');
    }).catchError((err, stack) {
      debugPrint('[HayaatiAlarm][DIAG] Alarm.init() a échoué : $err\n$stack');
    });
    control.addInvokeMethodListener(_onInvokeMethod);
    debugPrint(
      '[HayaatiAlarm][DIAG] addInvokeMethodListener enregistré pour '
      'instance=${identityHashCode(this)}',
    );
  }

  @override
  void dispose() {
    // ⚠️ DIAGNOSTIC : si cette ligne apparaît peu après init(), le service
    // est détruit avant que Python n'ait eu le temps d'appeler quoi que
    // ce soit, ce qui expliquerait un timeout systématique.
    debugPrint(
      '[HayaatiAlarm][DIAG] dispose() — instance=${identityHashCode(this)}',
    );
    control.removeInvokeMethodListener(_onInvokeMethod);
    super.dispose();
  }

  /// Copie un asset embarqué dans le bundle Flutter vers un fichier réel
  /// sur disque, une seule fois, puis renvoie le chemin absolu.
  ///
  /// [assetKey] est la clé complète telle qu'elle existe dans
  /// flutter_assets, par exemple :
  /// "packages/hayaati_alarm/assets/sounds/adhan.mp3"
  ///
  /// Le fichier est mis en cache dans le dossier de support de l'app.
  /// Au démarrage suivant, s'il existe déjà et n'est pas vide, on le
  /// réutilise directement sans relire le bundle.
  Future<String> _ensureAudioFileExtracted(String assetKey) async {
    final dir = await getApplicationSupportDirectory();
    final fileName = assetKey.split('/').last;
    final file = File('${dir.path}/$fileName');

    if (await file.exists() && await file.length() > 0) {
      return file.path;
    }

    final byteData = await rootBundle.load(assetKey);
    await file.writeAsBytes(
      byteData.buffer.asUint8List(
        byteData.offsetInBytes,
        byteData.lengthInBytes,
      ),
      flush: true,
    );
    debugPrint(
      '[HayaatiAlarm][DIAG] Audio extrait vers ${file.path} '
      '(${await file.length()} octets)',
    );
    return file.path;
  }

  Future<dynamic> _onInvokeMethod(String name, dynamic args) async {
    // ⚠️ DIAGNOSTIC : première ligne, inconditionnelle, avant tout switch
    // ou try. Si cette trace n'apparaît jamais dans le logcat pendant
    // qu'un timeout se produit côté Python, la fonction n'est simplement
    // jamais appelée : le problème est dans l'enregistrement du listener,
    // pas dans la logique métier ci-dessous.
    debugPrint(
      '[HayaatiAlarm][DIAG] _onInvokeMethod APPELÉ — name=$name '
      'instance=${identityHashCode(this)}',
    );
    try {
      switch (name) {
      case "setAlarm":
        final double volume = (args["volume"] as num?)?.toDouble() ?? 1.0;
        final num? fadeDurationSeconds = args["fadeDuration"] as num?;

        final volumeSettings = fadeDurationSeconds != null
            ? VolumeSettings.fade(
                volume: volume,
                fadeDuration: Duration(
                  milliseconds: (fadeDurationSeconds.toDouble() * 1000).round(),
                ),
              )
            : VolumeSettings.fixed(volume: volume);

        // ⚠️ CORRECTION 08/09/2026 : args["assetAudioPath"] contient la
        // clé d'asset brute envoyée par Python, pas un chemin de fichier
        // réel. On extrait le fichier nous-mêmes avant de le transmettre
        // à AlarmSettings, plutôt que de laisser le paquet 'alarm' tenter
        // (et rater) cette extraction de son côté.
        final resolvedAudioPath = await _ensureAudioFileExtracted(
          args["assetAudioPath"] as String,
        );

        final settings = AlarmSettings(
          id: args["id"] as int,
          dateTime: DateTime.parse(args["dateTime"] as String),
          assetAudioPath: resolvedAudioPath,
          loopAudio: args["loopAudio"] as bool? ?? true,
          vibrate: args["vibrate"] as bool? ?? true,
          androidFullScreenIntent:
              args["androidFullScreenIntent"] as bool? ?? true,
          volumeSettings: volumeSettings,
          warningNotificationOnKill:
              args["warningNotificationOnKill"] as bool? ?? false,
          notificationSettings: NotificationSettings(
            title: args["title"] as String? ?? "HAYAATI",
            body: args["body"] as String? ?? "",
            stopButton: args["stopButton"] as String? ?? "Arrêter",
          ),
        );
        await Alarm.set(alarmSettings: settings);
        return true;

      case "cancelAlarm":
        await Alarm.stop(args["id"] as int);
        return true;

      case "stopAll":
        await Alarm.stopAll();
        return true;

      case "requestPermissions":
        // POST_NOTIFICATIONS (Android 13+). Vérifié via permission_handler,
        // package déjà associé au paquet 'alarm' dans la documentation
        // de la communauté pour ce cas d'usage précis.
        final notifStatus = await Permission.notification.status;
        if (notifStatus.isDenied) {
          await Permission.notification.request();
        }
        return true;

      case "requestExactAlarmPermission":
        // SCHEDULE_EXACT_ALARM (Android 12+).
        final exactStatus = await Permission.scheduleExactAlarm.status;
        if (exactStatus.isDenied) {
          await Permission.scheduleExactAlarm.request();
        }
        return true;

      case "requestFullScreenIntentPermission":
        // ⚠️ Non implémenté : permission_handler ne couvre pas
        // USE_FULL_SCREEN_INTENT (demande fermée "not planned" par les
        // mainteneurs). Une vraie implémentation demanderait du code
        // natif Android (Kotlin) appelant
        // Settings.ACTION_MANAGE_APP_USE_FULL_SCREEN_INTENT, non écrit
        // pour l'instant. Les apps avec fonction d'alarme reçoivent
        // normalement cette permission par défaut à l'installation.
        // ignore: avoid_print
        print(
          '[HayaatiAlarm] requestFullScreenIntentPermission : non implémenté '
          '(voir commentaire dans hayaati_alarm.py). Vérifier manuellement '
          'sur le téléphone : Réglages > Apps > Hayaati > Autorisations spéciales.',
        );
        return true;

      default:
        return null;
      }
    } catch (err, stack) {
      // ⚠️ DIAGNOSTIC : sans ce catch, une exception ici (par exemple dans
      // Alarm.set() ou permission_handler) disparaissait silencieusement
      // côté Dart, sans jamais renvoyer d'erreur à Python — d'où le
      // timeout de 10s côté Python plutôt qu'un message d'échec net.
      debugPrint('[HayaatiAlarm][DIAG] Exception dans $name : $err\n$stack');
      rethrow;
    }
  }
}

File: hayaati_alarm/hayaati_alarm/src/hayaati_alarm/hayaati_alarm.py
"""
hayaati_alarm/hayaati_alarm.py

Service Flet pour un vrai réveil natif (son en boucle, plein écran,
volume forcé, bouton d'arrêt), construit sur le paquet Flutter mature et
maintenu 'alarm' (gdelataillade/alarm sur pub.dev), plutôt que sur du code
Android JNI/pyjnius écrit à la main.

Toute la communication passe par le protocole Flet standard
(invoke_method), exactement le même mécanisme que flet-android-notifications
utilise déjà avec succès dans ce projet. Aucun jnius, aucun autoclass,
aucun risque de crash JNI natif : une erreur ici redevient une Exception
Python normale, interceptable.

⚠️ CONFIRMÉ : la méthode d'appel héritée de ft.Service est bien
_invoke_method (avec underscore), vérifié par
python -c "import flet as ft; print([m for m in dir(ft.Service) if 'invoke' in m.lower()])"
qui renvoie ['_invoke_method'].

── Journal des corrections (30/08/2026) ─────────────────────────────
v1.1 :
  • Ajout de fade_duration (fondu sonore progressif, en secondes) et de
    warning_notification_on_kill (notification si l'utilisateur tue
    l'appli avant que l'alarme sonne). Ces deux options existent côté
    Dart (paquet alarm v5.1.5) mais n'étaient pas exposées ici, ce qui
    obligeait notification_scheduler.py à appeler des noms de paramètres
    qui n'existaient pas.
  • Suppression de toute référence à is_alarm_active() : cette méthode
    n'a jamais fonctionné, _invoke_method() ne remonte pas de valeur de
    façon synchrone dans ce projet. La prévention des doublons est déjà
    assurée par cancel_all_prayer_alarms(), appelée avant chaque
    reschedule_all() côté notification_scheduler.py.
  • bypass_dnd n'a pas été repris : aucune trace confirmée d'un tel
    paramètre dans l'API du paquet 'alarm'. Le contournement du mode
    Ne pas déranger est en pratique déjà couvert par
    androidFullScreenIntent, qui s'appuie sur AlarmManager en catégorie
    alarme côté Android.
─────────────────────────────────────────────────────────────────────

── Vérification de cohérence (09/09/2026) ────────────────────────────
Aucun changement fonctionnel : les noms de méthodes et de paramètres
restent exactement alignés avec core/notification_scheduler.py et
hayaati_alarm.dart. Seule correction : les deux exemples de chemin
d'asset dans les docstrings ci-dessous montraient encore
'assets/sounds/adhan.mp3' sans le préfixe 'packages/hayaati_alarm/',
alors que ce préfixe s'est révélé nécessaire (voir hayaati_alarm.dart,
_ensureAudioFileExtracted). Un exemple resté périmé aurait pu
réintroduire ce bug si quelqu'un le recopiait tel quel.
─────────────────────────────────────────────────────────────────────
"""
from __future__ import annotations
import flet as ft


@ft.control("HayaatiAlarm")
class HayaatiAlarm(ft.Service):
    """
    Réveil natif avec son en boucle et écran plein écran.

    Utilisation :
        alarme = HayaatiAlarm()
        page.services.append(alarme)  # ou l'équivalent d'enregistrement
                                        # de service utilisé ailleurs dans
                                        # le projet pour flet-android-notifications

        alarme.set_alarm(
            id=1010,
            date_time_iso="2026-08-25T05:12:00",
            asset_audio_path="packages/hayaati_alarm/assets/sounds/adhan_fajr.mp3",
            title="🕌 Fajr",
            body="C'est l'heure de la prière Fajr.",
        )
    """

    async def set_alarm(
        self,
        id: int,
        date_time_iso: str,
        asset_audio_path: str,
        title: str,
        body: str,
        loop_audio: bool = True,
        vibrate: bool = True,
        full_screen_intent: bool = True,
        volume: float = 1.0,
        fade_duration: float | None = None,
        warning_notification_on_kill: bool = False,
        stop_button: str = "Arrêter",
    ) -> None:
        """Programme une alarme native.

        asset_audio_path est la clé d'asset Flutter COMPLÈTE, avec le
        préfixe du package (ex. 'packages/hayaati_alarm/assets/sounds/adhan.mp3'),
        pas un chemin relatif type 'assets/sounds/adhan.mp3'. Confirmé le
        08/09/2026 en inspectant l'APK décompressé (voir
        core/notification_scheduler.py, DEFAULT_PRAYER_PREFS) : sans le
        préfixe packages/hayaati_alarm/, hayaati_alarm.dart échoue
        silencieusement à extraire le fichier audio du bundle.

        fade_duration : durée du fondu sonore progressif, en secondes.
        Laisser à None pour un volume fixe dès le début de l'alarme
        (comportement d'origine, inchangé par défaut).

        ⚠️ 31/08→01/09/2026 : _invoke_method() s'est révélée être une
        coroutine dans cette version de Flet (confirmé par le
        RuntimeWarning 'coroutine ... was never awaited' vu en logcat).
        Toutes les méthodes de ce fichier sont donc désormais async et
        utilisent await ; tous les appelants (notification_scheduler.py,
        main.py) ont dû être mis à jour en cascade.
        """
        await self._invoke_method(
            "setAlarm",
            {
                "id": id,
                "dateTime": date_time_iso,
                "assetAudioPath": asset_audio_path,
                "loopAudio": loop_audio,
                "vibrate": vibrate,
                "androidFullScreenIntent": full_screen_intent,
                "volume": volume,
                "fadeDuration": fade_duration,
                "warningNotificationOnKill": warning_notification_on_kill,
                "title": title,
                "body": body,
                "stopButton": stop_button,
            },
        )

    async def cancel_alarm(self, id: int) -> None:
        """Annule une alarme précise par son identifiant."""
        await self._invoke_method("cancelAlarm", {"id": id})

    async def stop_all_alarms(self) -> None:
        """Coupe toutes les alarmes en cours et à venir (ex. avant une
        replanification complète, comme reschedule_all())."""
        await self._invoke_method("stopAll", {})

    async def request_permissions(self) -> None:
        """Demande la permission de notification de base (POST_NOTIFICATIONS
        sur Android 13+), via permission_handler côté Dart."""
        await self._invoke_method("requestPermissions", {})

    async def request_exact_alarm_permission(self) -> None:
        """Demande la permission d'alarme exacte (SCHEDULE_EXACT_ALARM sur
        Android 12+), via permission_handler côté Dart. Sans cette
        permission accordée, Android peut retarder ou ignorer une alarme
        programmée avec Alarm.set(), sans erreur visible."""
        await self._invoke_method("requestExactAlarmPermission", {})

    async def request_full_screen_intent_permission(self) -> None:
        """Vérifie/journalise l'état de la permission plein écran
        (USE_FULL_SCREEN_INTENT sur Android 14+).

        ⚠️ Non implémenté en profondeur pour l'instant : permission_handler
        ne propose pas cette permission (demande de fonctionnalité fermée
        "not planned" par les mainteneurs). La demander correctement
        nécessiterait du code natif Android (Kotlin) dans l'extension,
        ce qui n'a pas été fait. Les applications ayant une vraie fonction
        d'alarme, comme HAYAATI, reçoivent normalement cette permission
        accordée par défaut à l'installation (contrairement aux
        applications qui en abusent pour des publicités plein écran) :
        https://source.android.com/docs/core/permissions/fsi-limits
        Cet appel journalise simplement l'intention côté Dart pour
        l'instant ; à vérifier sur le terrain avant d'investir dans le
        code natif si un test réel montre que ce n'est pas suffisant."""
        await self._invoke_method("requestFullScreenIntentPermission", {})
